mozjs/src/jsimpls.rs

```diff
@@ -380,20 +381,117 @@ impl<T> JS::Rooted<T> {
 }
 
 impl JS::ObjectOpResult {
+    pub fn ok(&self) -> bool {
+        assert_ne!(self.code_, JS::ObjectOpResult_SpecialCodes::Uninitialized as usize);
+        self.code_ == JS::ObjectOpResult_SpecialCodes::OkCode as usize
+    }
+
     /// Set this ObjectOpResult to true and return true.
     pub fn succeed(&mut self) -> bool {
         self.code_ = JS::ObjectOpResult_SpecialCodes::OkCode as usize;
         true
     }
 
+    pub fn fail(&mut self, code: JSErrNum) -> bool {
+        assert_ne!(self.code_, JS::ObjectOpResult_SpecialCodes::OkCode as usize);
+        self.code_ = code as usize;
```

Comment:
This does something different than the C++ API which asserts that the argument isn't JS::ObjectOpResult_SpecialCodes::OkCode. Here you are asserting that the existing code isn't that.

From https://searchfox.org/mozilla-central/source/js/public/Class.h#123:
```c++
  bool fail(uint32_t msg) {
    MOZ_ASSERT(msg != OkCode);
    code_ = msg;
    return true;
  }
```